src/hotspot/share/code/nmethod.cpp

```diff
@@ -1393,6 +1393,233 @@ nmethod::nmethod(
+
+  // Copy all nmethod data outside of header
+  memcpy(content_begin(), nm.content_begin(), nm.size() - nm.header_size());
+
+  RelocIterator iter(this);
+  CodeBuffer src((CodeBlob *)&nm);
+  CodeBuffer dst(this);
+  while (iter.next()) {
+    iter.reloc()->fix_relocation_after_move(&src, &dst);
+  }
+
+  ICache::invalidate_range(code_begin(), code_size());
+
+  _method                       = nm._method;
+
+  post_init();
+
+  // Update corresponding Java method to point to this nmethod
+  MutexLocker ml(NMethodState_lock, Mutex::_no_safepoint_check_flag);
+  if (_method != nullptr && _method->code() == &nm) {
+    methodHandle mh(Thread::current(), _method);
+    _method->set_code(mh, this, true);
+  }
+}
+
+nmethod* nmethod::relocate_to(nmethod* nm, CodeBlobType code_blob_type) {
+  if (nm == nullptr || !nm->is_relocatable()) {
+    return nullptr;
+  }
+
+  // No need to relocate if already in correct code heap
+  if (nm->lookup_code_blob_type() == code_blob_type) {
+    return nm;
+  }
+
+  nmethod* nm_copy = nullptr;
+
+  {
+    // Clear inline caches before acquiring any locks
+    VM_ClearNMethodICs clear_nmethod_ics(nm);
+    VMThread::execute(&clear_nmethod_ics);
+
+    MutexLocker mu(CodeCache_lock, Mutex::_no_safepoint_check_flag);
+    nm_copy = new (nm->size(), code_blob_type) nmethod(*nm);
+
+    if (nm_copy != nullptr) {
+      // To make dependency checking during class loading fast, record
+      // the nmethod dependencies in the classes it is dependent on.
+      // This allows the dependency checking code to simply walk the
+      // class hierarchy above the loaded class, checking only nmethods
+      // which are dependent on those classes.  The slow way is to
+      // check every nmethod for dependencies which makes it linear in
+      // the number of methods compiled.  For applications with a lot
+      // classes the slow way is too slow.
+      for (Dependencies::DepStream deps(nm_copy); deps.next(); ) {
+        if (deps.type() == Dependencies::call_site_target_value) {
+          // CallSite dependencies are managed on per-CallSite instance basis.
+          oop call_site = deps.argument_oop(0);
+          MethodHandles::add_dependent_nmethod(call_site, nm_copy);
+        } else {
+          InstanceKlass* ik = deps.context_type();
+          if (ik == nullptr) {
+            continue;  // ignore things like evol_method
+          }
+          // record this nmethod as dependent on this klass
+          ik->add_dependent_nmethod(nm_copy);
+        }
+      }
+    }
+  }
+  // Do verification and logging outside CodeCache_lock.
+  if (nm_copy != nullptr) {
+    NOT_PRODUCT(note_java_nmethod(nm_copy));
+    // Safepoints in nmethod::verify aren't allowed because nm_copy hasn't been installed yet.
+    DEBUG_ONLY(nm_copy->verify();)
+    nm_copy->log_new_nmethod();
+  }
+
+  nm->make_not_used();
+
+  return nm_copy;
+}
+
+bool nmethod::is_relocatable() const {
+  if (is_not_entrant()) {
+    return false;
+  }
+
```

Comment:
We plant to relocate only compiled code. We can check this with:
```
if (compiler_type() == compiler_none) {
  return false;
}
```

I think we don't need `is_method_handle_intrinsic()` and `is_continuation_native_intrinsic()` with this check.